# MetaBinning
[Snakemake](https://snakemake.readthedocs.io/en/stable/getting_started/installation.html) workflow for constructing MAGs from Illumina, PacBio, or ONT clean reads and assemblies.

The user provides the paths to clean Illumina, ONT, or PacBio reads and metagenomics contigs and/or scaffolds. This workflow runs `concoct`, `semibin2`, `MaxBin2`, and `MetaBat2` to construct MAGs. This workflow eliminates intermediate files for efficiency purposes. The output of this workflow is MAGs generated by multiple binning tools, along with their corresponding completeness/contamination statistics. 

## Installation

1. Install [conda](https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html).
   
3. Create a conda environment
   
<pre><code>conda create -n snakemake-7.3.7 python=3.10 -y 
conda activate snakemake-7.3.7</code></pre> 

3. Install snakemake v7.3.7 (using conda or mamba, a faster conda alternative)

<pre><code>conda create -n snakemake-7.3.7 -c conda-forge -c bioconda snakemake=7.3.7</code></pre>

If you don't have access to bioconda, you can also try installing this using `pip`

<pre><code>pip install snakemake==7.3.7</code></pre>

4. Clone the repository

<pre><code>git clone https://github.com/luisgv467/MetaBinning_short-reads.git</code></pre>

5. Authorize the execution of scripts:

This workflow needs to run scripts to execute concoct properly; however, it needs execution permission to work:
<pre><code> chmod +x scripts/* </code></pre>

## Run snakemake workflow

1. Get the basename of the samples you intend to process. You can do this by listing your samples available in your assembly directory:

<pre><code>ls /data/pam/lg21g/scratch/chapters/chapter_1/Resp_microbiome_db/Assembly/*.fa | sed -e 's|/data/pam/lg21g/scratch/chapters/chapter_1/Resp_microbiome_db/Assembly/||' -e 's|.fa||' > metagenome_paths.txt</code></pre>

2. Modify the `config/config.yml` file with your input and output locations. This workflow assumes that your metagenomic assemblies file extension is `.fa` and that your clean Illumina reads have a `_1.fastq.gz` and `_2.fastq.gz` file extension. Also, make sure to unmark which type of reads (Illumina, ONT, or PacBio) you plan to use. This workflow assumes that if you choose to work with Illumina, your input is clean paired reads. Otherwise, the protocol will assume you have an input of unpaired reads with a `.fastq.gz` extension. 

<pre><code># Required arguments

#List of input files
input:
    "/data/pam/lg21g/scratch/Programs/snakemake/MetaBinning/metagenome_paths.txt"

#Path to assemblies
assembly:
    "/data/pam/lg21g/scratch/chapters/chapter_1/Resp_microbiome_db/Assembly-long/Polished/"

#Path to clean reads
reads:
    "/data/pam/lg21g/scratch/chapters/chapter_1/Resp_microbiome_db/clean_long_reads/"

#Path to store alignments
alignment:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning-ONT/Alignment/"

#Path to store 10k-contigs
10k-contigs:
   "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning-ONT/10K-contigs/"

#Path to store metabat2 results
metabat2:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning-ONT/MetaBat2/"

#Path to store concoct results
concoct:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning-ONT/Concoct/"

concoct-final:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning-ONT/Concoct-final/"

#Path to store maxbin2 results
maxbin2:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning-ONT/MaxBin2/"

#Path to store semibin2 results
semibin2:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning-ONT/SemiBin2/"

#Final output of the bins
output:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning-ONT/All_bins/"

#checkM output
checkM:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning-ONT/CheckM/"

#Sequencing technology: choose one of Illumina, ONT, PacBio
tech: "Illumina"
#tech: "ONT"
#tech: "PacBio" </pre></code>

3. Run Snakemake

If using a server:
<pre><code>snakemake --use-conda -k -j 100 --profile config/lfs --rerun-incomplete --latency-wait 120 --scheduler greedy </pre></code>

If running locally:
<pre><code>snakemake --use-conda -k -j 10 --rerun-incomplete --latency-wait 60 </pre></code>

## Output

The MAGs produced by multiple binning tools will be stored in the directory `Binning/All_bins`. 
