# MetaBinning
[Snakemake](https://snakemake.readthedocs.io/en/stable/getting_started/installation.html) workflow for constructing MAGs from Illumina, PacBio, or ONT clean reads and assemblies.

The user provides the paths to clean Illumina reads and metagenomics contigs and/or scaffolds. This workflow runs `concoct`, `semibin2`, `MaxBin2`, and `MetaBat2` to construct MAGs. This workflow eliminates intermediate files for efficiency purposes. The output of this workflow is MAGs generated by multiple binning tools, along with their corresponding completeness/contamination statistics. 

## Installation

1. Install [conda](https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html).
   
3. Create a conda environment
   
<pre><code>conda create -n snakemake-7.3.7 python=3.10 -y 
conda activate snakemake-7.3.7</code></pre> 

3. Install snakemake v7.3.7 (using conda or mamba, a faster conda alternative)

<pre><code>conda create -n snakemake-7.3.7 -c conda-forge -c bioconda snakemake=7.3.7</code></pre>

If you don't have access to bioconda, you can also try installing this using `pip`

<pre><code>pip install snakemake==7.3.7</code></pre>

4. Clone the repository

<pre><code>git clone https://github.com/luisgv467/MetaBinning_short-reads.git</code></pre>

5. Install `CheckM` (if it's not installed already)

By using conda: 
<pre><code>conda install bioconda::checkm-genome</code></pre>

If you don't have access to `anaconda` or `bioconda`, you can try installing checkM via `pip` and its dependencies via `conda-forge`:
<pre><code>#Install CheckM
pip install checkm-genome

#Install CheckM dependencies
conda install -c conda-forge hmmer prodigal pplacer</code></pre>

6. Download `CheckM` reference database:

<pre><code>#Download the database:
wget https://data.ace.uq.edu.au/public/CheckM_data_2015_01_16.tar.gz

#Unpack
tar -xvzf CheckM_data_2015_01_16.tar.gz</code></pre>

Then you need to indicate to CheckM where the database is by doing:

Either set it via the command:
<pre><code>checkm data setRoot /path/to/checkm_data</code></pre>

Or export an environment variable in your shell startup file (.bashrc, .zshrc, etc.):
<pre><code>export CHECKM_DATA_PATH=/path/to/checkm_data</code></pre>

7. Authorize the execution of scripts:

This workflow needs to run scripts to execute concoct properly; however, it needs execution permission to work:
<pre><code> chmod +x scripts/* </code></pre>

## Run snakemake workflow

1. Get the basename of the samples you intend to process. You can do this by listing your samples available in your assembly directory:

<pre><code>ls /data/pam/lg21g/scratch/chapters/chapter_1/Resp_microbiome_db/Assembly/*.fa | sed -e 's|/data/pam/lg21g/scratch/chapters/chapter_1/Resp_microbiome_db/Assembly/||' -e 's|.fa||' > metagenome_paths.txt</code></pre>

2. Modify the `config/config.yml` file with your input and output locations. This workflow assumes that your metagenomic assemblies file extension is `.fa` and that your clean Illumina reads have a `_1.fastq.gz` and `_2.fastq.gz` file extension.

<pre><code># Required arguments

# Required arguments

#List of input files (File with the basename of samples to process)
input:
    "/data/pam/lg21g/scratch/Programs/snakemake/MetaBinning/metagenome_paths.txt"

#Path to assemblies
assembly:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/Assembly/"

#Path to clean reads
reads:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/clean_paired_reads/"

#Path to store alignments
alignment:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning/Alignment/"

#Path to store 10k-contigs
10k-contigs:
   "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning/10K-contigs/"

#Path to store metabat2 results
metabat2:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning/MetaBat2/"

#Path to store concoct results
concoct:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning/Concoct/"

concoct-final:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning/Concoct-final/"

#Path to store maxbin2 results
maxbin2:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning/MaxBin2/"

#Path to store semibin2 results
semibin2:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning/SemiBin2/"

#Final output of the bins
output:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning/All_bins/"

#checkM output
checkM:
    "/data/pam/lg21g/scratch/random/PT_sputum_samples/Binning/CheckM/" </pre></code>

3. Run Snakemake

If using a server:
<pre><code>snakemake --use-conda -k -j 100 --profile config/lfs --rerun-incomplete --latency-wait 120 --scheduler greedy </pre></code>

If running locally:
<pre><code>snakemake --use-conda -k -j 10 --rerun-incomplete --latency-wait 60 </pre></code>

## Output

The MAGs produced by multiple binning tools will be stored in the directory `Binning/All_bins`. The `CheckM` report will be stored in the `Binning/CheckM` directory.

